internal/provider/webhook/resource_webhook.go

```diff
@@ -0,0 +1,530 @@
+	if state.Config == nil {
+		state.Config = &webhookConfig{}
+	}
+	state.Config.URL = types.StringValue(webhook.Config.URL)
+
+	webhookSubscriptions := make([]webhookSubscription, len(webhook.Subscriptions))
+	for i, subscription := range webhook.Subscriptions {
+		newSubscription := webhookSubscription{
+			Events: make([]webhookSubscriptionEvent, 0),
+		}
+
+		if subscription.ResourceID != "" {
+			newSubscription.ResourceID = types.StringValue(subscription.ResourceID)
+		}
+
+		eventsMap := make(map[types.String][]types.String)
+		for _, event := range subscription.Events {
+			eventsMap[types.StringValue(event.Source)] = append(eventsMap[types.StringValue(event.Source)],
+				types.StringValue(event.Action))
+		}
+
+		for source, actions := range eventsMap {
+			newSubscription.Events = append(newSubscription.Events, webhookSubscriptionEvent{
+				Actions: actions,
+				Source:  source,
+			})
+		}
+
+		webhookSubscriptions[i] = newSubscription
+	}
+	if len(webhook.Subscriptions) > 0 ||
+		len(webhook.Subscriptions) != len(state.Subscriptions) {
+		state.Subscriptions = webhookSubscriptions
+	}
+
+	// Save updated state into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+func (r *resourceWebhook) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var plan, state webhook
+
+	// Read Terraform plan and state into the models
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if !plan.Name.Equal(state.Name) {
+		updateNameParams := webhookservice.NewWebhookServiceUpdateWebhookNameParams()
+		updateNameParams.ResourceName = state.ResourceName.ValueString()
+		updateNameParams.Body = &webhookmodels.HashicorpCloudWebhookUpdateWebhookNameRequestBody{
+			Name: plan.Name.ValueString(),
+		}
+
+		res, err := r.client.Webhook.WebhookServiceUpdateWebhookName(updateNameParams, nil)
+		if err != nil {
+			resp.Diagnostics.AddError("Error updating webhook name", err.Error())
+			return
+		}
+
+		webhook := res.GetPayload().Webhook
+		if webhook == nil {
+			resp.Diagnostics.AddError(
+				"Unexpected service response",
+				"The update webhook name request didn't fail but returned a nil webhook object. "+
+					"Report this issue to the provider developers.")
+			return
+		}
+		state.ResourceName = types.StringValue(webhook.ResourceName)
+
+		defer func() {
+			if resp.Diagnostics.HasError() {
+				// If it failed to update the resource but the name still got successfully updated,
+				// we update the name and resource name in the state.
+				// The next apply the new name will no longer be an update.
+				state.Name = plan.Name
+				resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+			}
+		}()
+	}
+
+	var updateMaks []string
+	if !plan.Config.URL.Equal(state.Config.URL) ||
+		!plan.Config.HmacKey.Equal(state.Config.HmacKey) {
```

Comment:
I couldn't get the provider to detect that I had changed my hmac key

Reply:
I did test that but maybe I made a change in between that changed the behavior. I will have a look and an acceptance test for this too.

Reply:
I tested it again and I couldn't reproduce what you experienced. I added two new acceptance tests to test that it is possible to update from none to an hmac_key and from one hmac_key to another.  Can you try one more time?